### PR TITLE
TranslateIngress changes

### PR DESCRIPTION
**Pull request by nicksardo**
**Changes**
 - Stop gracefully falling back on system default backend if specified backend is not found - no longer need EventRecorder in translater
 - Return list of errors from TranslateIngress
 - ensureIngress will error if TranslateIngress hits any errors. `ToSvcPorts` iterates all ingresses (for now) and will continue to collect all known service ports
 - Deleted a lot of tests in controller_test.go. Some of them should be re-created in different packages (tests that check GCE resource config), and new tests should be made for the controller that assert controller logic such as garbage collection, k8s object lifecycle edge cases.
 - Re-wrote tests for controller and created tests for TranslateIngress.

### CODE REVIEW

File: pkg/utils/utils.go [hunk 1]
```diff
@@ -150,6 +152,19 @@ func trimFieldsEvenly(max int, fields ...string) []string {
 	return ret
 }
 
+// PrettyJson marshals an object in a human-friendly format.
+func PrettyJson(data interface{}) (string, error) {
```
Comment (rramkumar1) [inline pkg/utils/utils.go]: Maybe I'm missing something, but this doesn't seem to be used anywhere.

File: pkg/utils/gceurlmap.go [hunk 2]
```diff
@@ -39,7 +39,41 @@ type PathRule struct {
 
 // NewGCEURLMap returns an empty GCEURLMap
 func NewGCEURLMap() *GCEURLMap {
-	return &GCEURLMap{hostRules: make(map[string][]PathRule)}
+	return &GCEURLMap{HostRules: make(map[string][]PathRule)}
+}
+
+func EqualStructure(a, b *GCEURLMap) bool {
```
Comment (rramkumar1) [inline pkg/utils/gceurlmap.go]: Can we call this Compare() instead?

File: pkg/test/utils.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+// Backend returns an IngressBackend with the given service name/port.
+func Backend(name string, port intstr.IntOrString) *extensions.IngressBackend {
+	return &extensions.IngressBackend{
+		ServiceName: name,
+		ServicePort: port,
+	}
+}
+
+// DecodeIngress returns deserializes an Ingress from yaml.
```
Comment (rramkumar1) [inline pkg/test/utils.go]: remove "returns" from comment.

File: pkg/utils/gceurlmap.go [hunk 4]
```diff
@@ -26,9 +26,9 @@ import (
 //       2. All paths for a specific host are unique.
 //       3. Adding paths for a hostname replaces existing for that host.
 type GCEURLMap struct {
-	DefaultBackend ServicePort
+	DefaultBackend *ServicePort
 	// hostRules is a mapping from hostnames to path rules for that host.
-	hostRules map[string][]PathRule
+	HostRules map[string][]PathRule
```
Comment (rramkumar1) [inline pkg/utils/gceurlmap.go]: Any reason for exposing the host rules? I left them package private so that we never access it without using the public getter

File: pkg/utils/gceurlmap.go [hunk 5]
```diff
@@ -49,54 +83,71 @@ func NewGCEURLMap() *GCEURLMap {
 func (g *GCEURLMap) PutPathRulesForHost(hostname string, pathRules []PathRule) {
 	// Convert the path rules to a map to filter out two equal paths
 	// Note that the if two paths are equal, the one later in the list is the winner.
-	uniquePaths := make(map[string]PathRule)
-	for _, pathRule := range pathRules {
-		if _, ok := uniquePaths[pathRule.Path]; ok {
-			glog.V(4).Infof("Equal paths (%v) for host %v. Using backend %+v", pathRule.Path, hostname, pathRule.Backend)
+	seen := make(map[string]struct{})
```
Comment (rramkumar1) [inline pkg/utils/gceurlmap.go]: can the value for this just be a bool? Might read a little better.

File: pkg/controller/translator/translator_test.go [hunk 6]
```diff
@@ -68,6 +63,127 @@ func fakeTranslator(negEnabled bool) *Translator {
 	return gce
 }
 
+func ingressFromFile(filename string) *extensions.Ingress {
```
Comment (rramkumar1) [inline pkg/controller/translator/translator_test.go]: Nit: Can we move these helpers to the bottom of the file? I don't like them being the first thing I see when I open the file :)

File: pkg/controller/translator/translator_test.go [hunk 7]
```diff
@@ -68,6 +63,127 @@ func fakeTranslator(negEnabled bool) *Translator {
...
+		log.Fatal(err)
+	}
+	v := &utils.GCEURLMap{}
+	if err := json.Unmarshal(data, v); err != nil {
+		log.Fatal(err)
+	}
+	return v
+}
+
+func TestTranslateIngress(t *testing.T) {
```
Comment (rramkumar1) [inline pkg/controller/translator/translator_test.go]: Nice job on this test, looks awesome. One thing I would like to see is maybe a case that results in 2 or more errors.

File: pkg/controller/controller_test.go [hunk 8]
```diff
@@ -163,345 +100,99 @@ func getKey(ing *extensions.Ingress, t *testing.T) string {
...
-		}
-	}
-	lbc.client.Extensions().Ingresses(ing.Namespace).Create(ing)
-	lbc.ctx.IngressInformer.GetIndexer().Add(ing)
-}
-
-func TestLbCreateDelete(t *testing.T) {
-	testFirewallName := "quux"
-	cm := NewFakeClusterManager(flags.DefaultClusterUID, testFirewallName)
+func TestIngressSyncError(t *testing.T) {
```
Comment (rramkumar1) [inline pkg/controller/controller_test.go]: Can we add a small comment to this test that is a little more descriptive than the function name itself? Same for other tests as well.

File: pkg/utils/utils.go (hunk 1 above)
Comment (nicksardo) [reply pkg/utils/utils.go]: Yea, but is a useful func to exist when we need to add more tests.

File: pkg/utils/gceurlmap.go (hunk 2 above)
Comment (nicksardo) [reply pkg/utils/gceurlmap.go]: It's only checking ServiceProjectID and `Compare` makes it sound like it's a full ServicePort check.

File: pkg/test/utils.go (hunk 3 above)
Comment (nicksardo) [reply pkg/test/utils.go]: Thanks

File: pkg/utils/gceurlmap.go (hunk 4 above)
Comment (nicksardo) [reply pkg/utils/gceurlmap.go]: Didn't have an immediate need. Was a little confused why `AllRules()` is used over `HostRules`. I have no strong opinion on this... will make it private

Comment (nicksardo) [reply pkg/utils/gceurlmap.go]: Ah, actually, we need exported for serializing/deserializing into JSON.

Comment (rramkumar1) [reply pkg/utils/gceurlmap.go]: Damn, okay. Let's get rid of AllRules() then too and just replace that with direct access.

File: pkg/controller/translator/translator_test.go (hunk 7 above)
Comment (nicksardo) [reply pkg/controller/translator/translator_test.go]: SG

File: pkg/utils/gceurlmap.go (hunk 4 above)
Comment (rramkumar1) [reply pkg/utils/gceurlmap.go]: Damn, okay. Let's get rid of AllRules() then too and just replace that with direct access.

Comment (nicksardo) [reply pkg/utils/gceurlmap.go]: Done
